### DISCUSSION TITLE
Slowly response

### DISCUSSION THREAD

**Discussion by EnricGarrigo** [jasonacox/tinytuya#250] [Q&A] [answered]
Hi all.
After some tests with Tuya devices, we have observed that once we have a load connected and powered, the API returns '0' to current values at least during 20-30 seconds. After this period, the 'cur_current' data is received with the right value.

Because that, if the load is powered on for a period less than 20 seconds (for example, in little intervals of 10 seconds), we have an accumulated current consumption of '0'.

Is it possible to accelerate the response time? Is any way to obtain the current values immediatly?

Thanks!

Update: if we enter in SmartLife app to see the values, the API refresh immediately. Any refresh on app provoke an immediately refresh of the values read from the API.

**Comment by jasonacox** [comment, marked as answer]
Hi @EnricGarrigo - it sounds like your device needs to the UPDATEDPS command to cause it to update the power metrics:

```python
import tinytuya
import time

d = tinytuya.OutletDevice(DEVICEID, DEVICEIP, DEVICEKEY)
d.set_version(3.1)

data = d.status()
print(data)

print(" > Wait 5 sec < ")
time.sleep(5)

print(" > Request Update < ")
result = d.updatedps()
# some require params that you wish updated
#result = d.updatedps(['18','19','20'])
print(result)

print(" > Fetch Status Again < ")
data2 = d.status()
print(data2)

print("")
print("Before %r" % data)
print("After  %r" % data2)

```

**Reply by EnricGarrigo** [reply]
Hi, Jason. Thanks for your answer.

I'm using the Cloud functions, because the PC with python and the device are not in the same network, but both have internet connection.
With cloud functions I don't know how to use the 'updatedps' function because I need an device object ('d' in your answer) that I don't have. I will appreciate your instructions about how to do that.
I'm sending the dps values throug mqtt protocol succesfully, but the values are 30 seconds delayed. This is my python code:
```python
# coding=utf-8
import tinytuya
import json
import paho.mqtt.client as mqtt
import time
import threading

mqtt_server = "localhost"
mqtt_port = 1883

my_id1 = "bf0339cbee33993ed5w3ox"

def read_publish_data():
	result = c.getstatus(my_id1)
	dps = result["result"]
	json_string = '{"id": ' +  '"' + my_id1 + '"'
	for my_dp in dps:
		json_element = '"' + my_dp["code"] + '": ' + '"' + str(my_dp["value"]) + '"'
		json_string = json_string + ", " + json_element
	json_string = json_string + "}"
	client.publish("iot/input/"+my_id1, json_string)

def subscriber():
	client.loop_forever()

def publisher():
	while True:
		time.sleep(0.5)
		read_publish_data()

def on_connect(client, userdata, flags, rc):
	if ( rc == 0):
		print("connected")
		client.subscribe("iot/output")
	else:
		print ("Connection Error", rc)

def on_subscribe(client, userdata, mid, granted_qos):
	print("subscribed") 

def on_message(client, userdata, msg):

	my_message = msg.payload.decode("utf-8")
	data = json.loads(my_message)

	my_id = data["device_id"]
	my_order = data["orden"]
	my_value = json.loads(data["parametro"].lower())

	if my_order == "SET":

		commands = {
		'commands': [{
		'code': 'switch',
		'value': my_value
		}]
		}
		print("Sending command...", commands)
		result = c.sendcommand(my_id,commands)
		print("Sended")

	if my_order == "READ":
		print("READ")
		rad_data(my_id)


c = tinytuya.Cloud()
devices = c.getdevices()


client = mqtt.Client()
client.on_connect = on_connect
client.on_message = on_message
client.on_subscribe = on_subscribe


client.connect(mqtt_server, mqtt_port)

sub = threading.Thread(target = subscriber)
pub = threading.Thread(target = publisher)

sub.start()
pub.start()
 

```

**Reply by jasonacox** [reply]
Hi @EnricGarrigo - Unless you are willing to pay Tuya for the extended use of their APIs, you are going to hit their limit to "free calls".  Part of the reason for TinyTuya was to be able to get local access and control of these Tuya devices. 

If you are not concerned about the API limits on the cloud, you can research the Cloud APIs here: https://developer.tuya.com/en/docs/cloud/device-connection-service?id=Kb0b8geg6o761 - I don't know how you request the equivalent of an UPDATEDPS via cloud.  However, I suggest trying to send commands to the device (e.g. turn on) and see if that results in an updated list of DPSs.

**Reply by EnricGarrigo** [reply]
Hi, jasonacox.

I expected to manage the Tuya devices through a relatively simple python code. If I have to use http requests, the script becomes more complex that I expected but, If that is the only way to refresh dps rapidly, we will have to do. We will try your suggestion about send command and if it doesn't work, migrate to http requests.
Thank you for your development and support.

**Reply by uzlonewolf** [reply]
> If I have to use http requests, the script becomes more complex that I expected

How so?  TinyTuya does all the heavy lifting for you:
```python
c = tinytuya.Cloud()
response = c.cloudrequest( '/some/request', query={'a':1, 'b':2} )
```

> the PC with python and the device are not in the same network

Can the PC "see" the device, or is the device behind a firewall?  As long as the PC can see the device you can still talk to it directly even if they're not on the same network.

**Reply by EnricGarrigo** [reply]
"How so?"

Because I guess that I need to generate all the authentication potocol instead of write the credentials on a simple config file.
With a simple call like your code, I obtain: "1008 | invalid access | The value of accessid is invalid."

I would appreciate if you can show me the full code with the mandatory authentication calls.

Many thanks.

**Reply by jasonacox** [reply]
HI @EnricGarrigo - That code assumes you have the tinytuya.json file already set up via the wizard. If you don't, you will need to provide your credentials like this to run the authentication calls for you:

```python
# Connect to Tuya Cloud
# c = tinytuya.Cloud()  # uses tinytuya.json 
c = tinytuya.Cloud(
        apiRegion="us", 
        apiKey="xxxxxxxxxxxxxxxxxxxx", 
        apiSecret="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", 
        apiDeviceID="xxxxxxxxxxxxxxxxxxID")

```

**Reply by EnricGarrigo** [reply]
Hi.
Following your appropriate advice, I have changed my python code for use tha API calls. For now, this is my part of code that reads the device dps:

while True:
    response = openapi.get("/v1.0/iot-03/devices/{}/status".format(DEVICE_ID))
    print(response)
    time.sleep(1)

When I run this code, the behaviour is the same that with tinytuya library: I'm getting a value of '0' for the 'cur_current' dps. If I switch on the radiator that is behind the smart switch, the value is still '0' during next 30 seconds. Only after this period, I receive a cur_current value of '8834', If I switch off the radiator, I receive a value of '0', five seconds after switch. If I switch on again, I receive a value of '8834' only 30 seconds after switch on.

With this behaviour, if I switch 'on' and 'off' the radiator every 10 seconds, I will obtain an accumulated value of '0' for the current, that is obviously not true.

**Reply by jasonacox** [reply]
Yes, that does not seem helpful.  It sounds like this device needs a "updateDPS" command. It may be that this is not possible using the cloud, I don't know. However, this can be done locally on your LAN.  I believe @uzlonewolf provided this suggestion which I would recommend as well:

> Can the PC "see" the device, or is the device behind a firewall? As long as the PC can see the device you can still talk to it directly even if they're not on the same network.

```python
import tinytuya
import time

d = tinytuya.OutletDevice(DEVICEID, DEVICEIP, DEVICEKEY)
d.set_version(3.1)
data = d.status()

print(" > Request Update < ")
result = d.updatedps()
# some require params that you wish updated
#result = d.updatedps(['18','19','20'])
print(result)

print(" > Fetch Status Again < ")
data2 = d.status()

print("")
print("Before %r" % data)
print("After update  %r" % data2)
```

**Reply by EnricGarrigo** [reply]
I don't have any IP for the device:

![Captura_2](https://user-images.githubusercontent.com/41141131/215080695-b1c20fb1-1350-49bd-8c6f-634cf176b8bf.PNG)

How can I use the 'local' version of tinytuya in this case?

Thanks

**Reply by jasonacox** [reply]
Do you know the network address of the LAN that runs the devices?  If it is 192.168.0.x, use something like this:

```bash
python3 -m tinytuya scan -force 192.168.0.0/24
```
